JRuby should provide the line number of errors when executing in the JSR223 Context

**Environment Information**
  
Provide at least:
* JRuby version:  9.2.14.0
* Operating system and platform:  Mac

**Expected Behavior**

When operating in a JSR223 environment if there is an evaluation error the line number should be provided in the error message.  This information currently exists in the stack trace.  However, It is common for systems using JSR223 to not print the full stack trace on evaluation error and just to print the exception message itself.  

This is consistent with the behavior of JavaScript, Groovy and Jython.  However, JRuby does not do this.

Here is an exception from Jython:
```
 Error during evaluation of script 'file:/openhab/conf/automation/jsr223/python/personal/test.py': NameError: name 'asdf' is not defined in <script> at line number 32
```

**Actual Behavior**

This is the same type of error, from JRuby:
```
Error during evaluation of script 'file:/openhab/conf/automation/jsr223/ruby/personal/test.rb': org.jruby.embed.EvalFailedException: (NameError) undefined local variable or method `asfd' for main:Object
```

It would be beneficial for JRuby to operate in a similarly as other JSR223 implementations.

Here is an example of an error inside a function in Jython:

```python
def test():
    asdf

test()
```

```
14:57:29.890 [ERROR] [ript.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/openhab/conf/automation/jsr223/python/personal/test.py': NameError: global name 'asdf' is not defined in <script> at line number 56
```

--- a/core/src/main/java/org/jruby/embed/jsr223/JRubyEngine.java
+++ b/core/src/main/java/org/jruby/embed/jsr223/JRubyEngine.java
@@ -47,6 +47,7 @@ import org.jruby.embed.ScriptingContainer;
 import org.jruby.exceptions.NoMethodError;
 import org.jruby.exceptions.RaiseException;
 import org.jruby.javasupport.JavaEmbedUtils;
+import org.jruby.runtime.backtrace.RubyStackTraceElement;
 import org.jruby.runtime.builtin.IRubyObject;
 
 /**
@@ -72,7 +73,7 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
         try {
             return new JRubyCompiledScript(container, this, script);
         } catch (RaiseException e) {
-            throw wrapException(e);
+            throw wrapRaiseException(e);
         }
     }
 
@@ -82,7 +83,7 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
         try {
             return new JRubyCompiledScript(container, this, reader);
         } catch (RaiseException e) {
-            throw wrapException(e);
+            throw wrapRaiseException(e);
         }
     }
 
@@ -95,6 +96,8 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
         try {
             IRubyObject ret = unit.get().run();
             return JavaEmbedUtils.rubyToJava(ret);
+        } catch (RaiseException e) {
+            throw wrapRaiseException(e);
         } catch (Exception e) {
             throw wrapException(e);
         } finally {
@@ -216,6 +219,8 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
             return container.callMethod(receiver, method, args, Object.class);
         } catch (NoMethodError e) {
             throw wrapMethodException(e);
+        } catch (RaiseException e) {
+            throw wrapRaiseException(e);
         } catch (Exception e) {
             throw wrapException(e);
         } finally {
@@ -235,6 +240,8 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
             return container.callMethod(receiver, method, args, Object.class);
         } catch (NoMethodError e) {
             throw wrapMethodException(e);
+        } catch (RaiseException e) {
+            throw wrapRaiseException(e);
         } catch (Exception e) {
             throw wrapException(e);
         } finally {
@@ -251,6 +258,25 @@ public class JRubyEngine implements Compilable, Invocable, ScriptEngine {
     }
 
     private static ScriptException wrapException(Exception e) {
+        return new ScriptException(e);
+    }
+
+    private static ScriptException wrapRaiseException(RaiseException e) {
+        RubyStackTraceElement[] trace = e.getException().getBacktraceElements();
+
+        if (trace.length > 0) {
+            RubyStackTraceElement top = trace[0];
+
+            String file = top.getFileName();
+            int line = top.getLineNumber();
+
+            if (file == null) file = "<script>";
+
+            ScriptException se = new ScriptException("Error during evaluation of Ruby in " + file + " at line " + line + ": " + e.getMessage());
+            se.initCause(e);
+            return se;
+        }
+
         return new ScriptException(e);
     }
 

--- a/core/src/test/java/org/jruby/embed/jsr223/JRubyEngineTest.java
+++ b/core/src/test/java/org/jruby/embed/jsr223/JRubyEngineTest.java
@@ -51,6 +51,7 @@ import org.jruby.embed.PositionFunction;
 import org.jruby.embed.RadioActiveDecay;
 import org.jruby.embed.internal.BiVariableMap;
 import org.jruby.exceptions.NoMethodError;
+import org.jruby.exceptions.RuntimeError;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
@@ -616,6 +617,25 @@ public class JRubyEngineTest extends BaseTest {
             fail("Expected javax.script.ScriptException");
         } catch (javax.script.ScriptException sex) {
             // javax.script.ScriptException is expected
+        }
+    }
+
+    /**
+     * Test of invokeMethod method throwing a Ruby exception.
+     */
+    @Test
+    public void testInvokeMethodWithRubyException() throws Exception {
+        ScriptEngine instance = newScriptEngine();
+        String filename = "scriptyscript.rb";
+        instance.put(ScriptEngine.FILENAME, filename);
+        instance.eval("def trigger_raise\nraise 'hello from ruby'\nend");
+        try {
+            ((Invocable) instance).invokeMethod(Object.class,"trigger_raise", null);
+            fail("Expected javax.script.ScriptException");
+        } catch (javax.script.ScriptException sex) {
+            String message = sex.getMessage();
+            assertTrue("message did not contain filename: \"" + message + "\"", message.contains(filename + " at line 2"));
+            assertEquals("cause was not RuntimeError: " + sex.getCause(), RuntimeError.class, sex.getCause().getClass());
         }
     }
 

